---
title: "Exposome Data Integration with Omic Data"
author: "Carles Hernandez-Ferer and Juan R. Gonzalez"
date: "`r doc_date()`"
package: "`r pkg_ver('rexposome')`"
abstract: >
  This is an introductory guide to integration analysis between exposome and omics data with R package rexposome. The document ilustrates two types of analysis: 1) Association analysis, that are performed between exposome and a single omic data-set; and 2) Integration analysis where multiple data-sets, including exposome data, are analysed at the same time.
vignette: >
  %\VignetteIndexEntry{Bioconductor style for HTML documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r setup, include=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(echo = TRUE, warnings=FALSE)
```

# Introduction

## Pipeline

## Data Format

## Installation

```{r github, eval=FALSE}
devtools::install_github("carleshf/rexposome", ref="devel")
```

The methods provided by `rexposome` and used in this document relies in other existing Bioconductor packages. All association and integration methods requires a `r Biocpkg("MultiDataSet")` object. The following are the R packages used in association analyses:

 * `r Biocpkg("snpStats")`
 * `r Biocpkg("limma")`
 * `r Biocpkg("MEAL")`

These are the packages used in the integration analyses:

 * `r Biocpkg("omicade4")`
 * `r CRANpkg("pma")`


# Analysis

__TEMP__: While `brgeDummyData` data package is not available, the data will be loaded using the current function:

```{r ldBRGE}
ldBRGE <- function(set) {
    path <- function(file) {
        prefix <- "C:/Users/carle/Desktop/rexposomeSupplementary/inma_sabadell"
        return(file.path(prefix, file))
    }
    sapply(set, function(nm) {
        pt <- NULL
        if(nm == "snp") {pt <- path("genome/snp.RData")}
        if(nm == "methy") {pt <- path("methylation/meth.RData")}
        if(nm == "gexp") {pt <- path("transcriptome/gexp_concordate_mset4.RData")}
        if(nm == "prot") {pt <- path("proteome/prot.RData")}
        
        load(pt, envir = .GlobalEnv)
    })
    invisible()
}
```

Under this sections two types of analysis are explained: __Association Analysis__ and __Integration Analysis__.

Both types of analysis requests a `MultiDataSet` object. While __Association Analysis__ are performed using a `MultiDataSet` object that contains an `ExposomeSet` plus a `SnpSet`, `MethylationSet` or `ExpressionSet`, __Integration Analysis__ needs a `MultiDataSet` object with, at last, two of these objects.

## Associaction Analysis

```{r load_mds, message=FALSE}
library(rexposome)
library(MultiDataSet)
```

### Exposome - Genome Data Associaction

The first step of an association analysis performed with `rexposome` is to create the `MultiDataSet` with the data to be analyzed.

The goal of this analysis is to perform a basic GWAS using the _phenotypes_ stored in an `ExposomeSet`. Hence, a `MultiDataSet` with an `ExposomeSet` and a `SnpSet` is needed.

```{r snp_data, cache=FALSE}
data("exposome", package = "rexposome")
ldBRGE("snp")
```

```{r snp_samples, message=FALSE, warning=FALSE}
sampleNames(snp4) <- gsub("id", "SIMSAB", sampleNames(snp4))
```

Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r snp_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_snps(mds, snp4)
mds <- add_exp(mds, expo)
```

After creating the `MultiDataSet` created the common samples between sets are kept and the others discarded:

```{r snp_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r snp_mds_show}
mds
```

The association tests is done using the method `assocSNP`. The arguments of this method are:

```{r snp_args}
args(assocSNP)
```

The `object` argument is filled with the `MultiDataSet` object created previously. The `design` is a formula where the dependent variable are the genotypes and the dependent the _heath outcome_. `family` argument must indicate the type of the _heath outcome_.

The following are the _health outcomes_ in the `ExposomeSet`:

```{r snp_phenotypes}
phenotypeNames(expo)
```

A GWAS for _asthma_ using `assocSNP` is done as:

```{r snp_analysis, message=FALSE, warning=FALSE}
gwas <- assocSNP(mds, asthma~sex+age)
```

The result is an object of class `ResultSet`:

```{r snp_show}
gwas
```

This object allows to plot the results and draw a QQ plot and a Manhattan plot:

```{r snp_qq}
plotAssociation(gwas, type="qq")
```

```{r snp_manhattan}
plotAssociation(gwas, type="manhattan")
```

The raw results obtained from GWAS can be extracted from the `ResultSet` with the method `raw`:

```{r snp_extract}
head(extract(gwas))
```

```{r snp_rm, echo=FALSE}
rm("expo", "snp4", "mds", "gwas")
```

### Exposome - Methylome Data Associaction

The goal of this analysis is to perform an EWAS between the methylation levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and a `MethylationSet` is required.

```{r methy_data, cache=FALSE}
data("exposome", package = "rexposome")
ldBRGE("methy")
```

```{r methy_samples, message=FALSE, warning=FALSE}
sampleNames(mset4) <- gsub("id", "SIMSAB", sampleNames(mset4))
```

Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r methy_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_methy(mds, mset4)
mds <- add_exp(mds, expo)
```

After creating the `MultiDataSet` created the common samples between sets are kept and the others discarded:

```{r methy_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r methy_mds_show}
mds
```

After obtaining the common samples, the association analysis between CpGs and exposures is done using `assocME`. This method requires a `MultiDataSet` object and a formula with the covariates. The optional argument `select` allows to perform the association on a sub-set of exposures (if `set = "exposures`) or sub-set of _outcomes_(if `set = "phenotypes"`).

```{r methy_analysis, message=FALSE, warning=FALSE}
ewas <- assocME(mds, ~sex+age, 
    select=c("logpb", "lcrCo_T1", "lTHM_T1", "Ben_preg", "garden_pesticides"))
```

The result is an object of class `ResultSet`:

```{r methy_show}
ewas
```

The raw results of the analysis can be obtained using the method `extract` and giving it the exposures of interest. If no exposure (or phenotype) is given, `extract` will return a table with the results for all the exposures (or phenotypes).

```{r methy_extract}
head(extract(ewas, rid="garden_pesticides"))
```

The `ResultSet` can be used to plot the already seen QQ and Manhattan plots. Also a volcano plot.

```{r methy_volcano, fig.width=6, fig.height=6}
plotAssociation(ewas, rid="garden_pesticides", type="volcano")
```

```{r methy_rm, echo=FALSE}
rm("expo", "mds", "ewas")
```

```{r message=FALSE, echo=FALSE, include=FALSE}
gc()
```

The same analysis can be done using an `ExposomeClust` instead of an `ExposomeSet`. 

```{r methy_cls, cache=FALSE}
data("eclust", package = "rexposome")
```

```{r methy_mds2, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_methy(mds, mset4)
mds <- add_cls(mds, expo_c)
```

After creating the `MultiDataSet` the method `assocME` can be called directly. It will look for the common samples before running the association test.

```{r methy_analysis2, message=FALSE, warning=FALSE}
methy <- assocME(mds, ~sex+age)
methy
```

The results of the association analysis can be obtained using the tag `"cluster"` in the argument `rid` of `extract` method:

```{r methy_extract2}
head(extract(methy, rid="cluster"))
```

And a Manhattan plot showing the association between the clustering and the protein levels of expression can be drawn:

```{r methy_manhattan2}
plotAssociation(methy, rid="cluster", type="volcano")
```

```{r methy_rm2, echo=FALSE}
rm("expo_c", "mset4", "mds", "methy")
```

### Exposome - Transcriptome Data Associaction

The aim of this analysis is to perform an association test between the gene expression levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and an `ExpressionSet` is required.

```{r gexp_data, cache=FALSE}
data("exposome", package = "rexposome")
ldBRGE("gexp")
```

```{r gexp_samples, message=FALSE, warning=FALSE}
sampleNames(gexp4) <- gsub("id", "SIMSAB", sampleNames(gexp4))
```

Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r gexp_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_genexp(mds, gexp4)
mds <- add_exp(mds, expo)
```

After creating the `MultiDataSet` created the common samples between sets are kept and the others discarded:

```{r gexp_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r gexp_mds_show}
mds
```

The association analysis between exposures and gene expression is done using the method `assocGE`. This method has the same behavior than `assocME`.

```{r gexp_analysis, message=FALSE, warning=FALSE}
gexp <- assocGE(mds, ~sex+age, 
    select=c("logpb", "lcrCo_T1", "lTHM_T1", "Ben_preg", "garden_pesticides"))
```

The obtained object is also a `ResultSet`. The raw table can be obtained using `extract`:

```{r gexp_extract}
tt <- extract(gexp)
table(tt$exposure)
head(tt[ , c("chromosome", "logFC", "P.Value", "adj.P.Val", "exposure")], n=5)
```

And all three QQ, Manhattan and volcano plots can be dawn using `plotAssociation`:

```{r gexp_volcano, fig.width=6, fig.height=6}
plotAssociation(gexp, rid="logpb", type="volcano")
```

```{r gexp_rm, echo=FALSE}
rm("expo", "mds", "gexp", "tt")
```

The same analysis can be done using an `ExposomeClust` instead of an `ExposomeSet`. 

```{r gexp_cls, cache=FALSE}
data("eclust", package = "rexposome")
```

```{r gexp_mds2, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_genexp(mds, gexp4)
mds <- add_cls(mds, expo_c)
```

After creating the `MultiDataSet` the method `assocGE` can be called directly. It will look for the common samples before running the association test.

```{r gexp_analysis2, message=FALSE, warning=FALSE}
gexp <- assocGE(mds, ~sex+age)
gexp
```

The results of the association analysis can be obtained using the tag `"cluster"` in the argument `rid` of `extract` method:

```{r gexp_extract2}
head(extract(gexp, rid="cluster"))
```

And a Manhattan plot showing the association between the clustering and the protein levels of expression can be drawn:

```{r gexp_manhattan2}
plotAssociation(gexp, rid="cluster", type="manhattan")
```

```{r gexp_rm2, echo=FALSE}
rm("expo_c", "gexp4", "mds", "gexp")
```

### Exposome - Proteome Data Associaction

The aim of this analysis is to perform an association test between the protein' expression levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and an `ExpressionSet` with protein expression data is required.

```{r prot_data, cache=FALSE}
data("exposome", package = "rexposome")
ldBRGE("prot")
```

```{r prot_cvrt, echo=FALSE}
colnames(prot4) <- gsub("_ok", "", colnames(prot4))
ad <- prot4
rownames(ad) <- ad$ID
ad <- ad[, -1]
pd <- data.frame(A=1:nrow(ad), B=1:nrow(ad), row.names=rownames(ad))
fd <- data.frame(A=1:ncol(ad), B=1:ncol(ad), row.names=colnames(ad))

prot4 <- ExpressionSet(
    assayData=assayDataNew("environment", exprs=t(ad)),
    featureData=AnnotatedDataFrame(fd), 
    phenoData=AnnotatedDataFrame(pd)
)

sampleNames(prot4) <- gsub("sam", "SIMSAB", sampleNames(prot4))

rm("ad", "pd", "fd")
```

Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r prot_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_prot(mds, prot4)
mds <- add_exp(mds, expo)
```

After creating the `MultiDataSet`, the common samples between sets are kept and the others discarded:

```{r prot_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r prot_mds_show}
mds
```

Ones obtained the common samples, the association analysis is done using `assocPRT`, that works like `assocME` and `assocGE`.

```{r prot_analysis, message=FALSE, warning=FALSE}
prot <- assocPRT(mds, ~sex+age)
prot
```

As seen before the result of `assocPRT` is a `ResultSet` object which results can be ploted using `plotAssociation`.

```{r prot_manhattan}
plotAssociation(prot, rid="ldde_lip", type="manhattan")
```

```{r prot_rm, echo=FALSE}
rm("expo", "mds", "prot")
```

The same analysis can be done using an `ExposomeClust` instead of an `ExposomeSet`. 

```{r prot_cls, cache=FALSE}
data("eclust", package = "rexposome")
```

```{r prot_mds2, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_prot(mds, prot4)
mds <- add_cls(mds, expo_c)
```

After creating the `MultiDataSet` the method `assocPRT` can be called directly. It will look for the common samples before running the association test.

```{r prot_analysis2, message=FALSE, warning=FALSE}
prot <- assocPRT(mds, ~sex+age)
prot
```

The results of the association analysis can be obtained using the tag `"cluster"` in the argument `rid` of `extract` method:

```{r prot_extract}
head(extract(prot, rid="cluster"))
```

And a Manhattan plot showing the association between the clustering and the protein levels of expression can be drawn:

```{r prot_manhattan2}
plotAssociation(prot, rid="cluster", type="manhattan")
```

```{r prot_rm2, echo=FALSE}
## rm("expo_c", "prot4", "mds", "prot")
rm("expo_c", "mds", "prot")
```

## Integration Analysis

The aim on the integration analysis is to analyze multiple data-sets in a single shot, looking for similar behaviors in terms of features and through the data-sets. `rexposome` allows to perform this analysis using two strategies: _multiple co-inertia analysis_ (`mcia`) and _sparse multiple canonical correlation analysis_ (`mcca`).

To perform an integration analysis using exposome, transcriptome and proteome data a `MultiDataSet` object is required with the thee data-sets.

```{r}
data("exposome", package = "rexposome")
ldBRGE("gexp")
## ldBRGE("prot")
```

```{r gexp_samples2, message=FALSE, warning=FALSE}
sampleNames(gexp4) <- gsub("id", "SIMSAB", sampleNames(gexp4))
```

All the data-sets needs to be complete. So the exposures should be imputed:

```{r impute_exposures, cache=TRUE}
expo <- impute(expo)
```

```{r crs_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_prot(mds, prot4)
mds <- add_genexp(mds, gexp4)
mds <- add_exp(mds, expo)
```

Once the data-sets is created, we look for the common samples.

```{r crs_mds_common, error=TRUE}
mds <- commonSamples(mds)
mds
```

Both integration analysis are done using the method `crossomics`. The argument `method` can takes the values `"mcia"` and `"mcca"`.

```{r crs_analysis1, warning=FALSE, message=FALSE}
crs1 <- crossomics(mds, method="mcia")
```

```{r crs_analysis2, warning=FALSE, message=FALSE}
crs2 <- crossomics(mds, method="mcca")
```

The method `crossomics` returns a `ResultSet` that can be used in the method `plotIntegration` to plot the results.


```{r crs_plot1, fig.width=12, fig.height=6}
plotIntegration(crs1)
```

```{r crs_plot2, fig.width=12, fig.height=10}
plotIntegration(crs2, lb.th=0.3)
```
